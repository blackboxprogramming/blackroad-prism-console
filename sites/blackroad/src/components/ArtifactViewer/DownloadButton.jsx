import React from 'react';

export default function DownloadButton({ url }) {
  return (
    <a
      href={url}
      target="_blank"
      rel="noreferrer"
      className="rounded border border-slate-300 bg-white px-3 py-1 text-xs font-medium text-slate-700 shadow hover:bg-slate-100"
    >
      Download
    </a>
  );
}
