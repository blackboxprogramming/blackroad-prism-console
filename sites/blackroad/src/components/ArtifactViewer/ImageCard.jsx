import React from 'react';

export default function ImageCard({ artifact }) {
  if (!artifact?.url) {
    return <div className="text-sm text-slate-500">Image unavailable.</div>;
  }
  return (
    <figure className="flex h-full w-full flex-col items-center justify-center gap-3">
      <img
        src={artifact.url}
        alt={artifact.name ?? 'Job artifact'}
        className="max-h-[480px] max-w-full rounded border border-slate-200 object-contain shadow"
      />
      {artifact.caption ? <figcaption className="text-xs text-slate-500">{artifact.caption}</figcaption> : null}
    </figure>
  );
}
