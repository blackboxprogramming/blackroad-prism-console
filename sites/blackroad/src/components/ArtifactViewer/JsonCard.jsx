import React from 'react';

export default function JsonCard({ artifact }) {
  if (!artifact) {
    return <div className="text-sm text-slate-500">No JSON payload.</div>;
  }
  const json = artifact.json ?? artifact.payload ?? artifact;
  return (
    <pre className="max-h-[520px] overflow-auto rounded border border-slate-200 bg-slate-900 p-4 text-xs text-slate-100">
      {JSON.stringify(json, null, 2)}
    </pre>
  );
}
