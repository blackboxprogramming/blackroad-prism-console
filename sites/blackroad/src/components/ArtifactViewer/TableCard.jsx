import React, { useMemo } from 'react';

function parseCsv(text = '') {
  return text
    .trim()
    .split(/\r?\n/)
    .map((row) => row.split(',').map((cell) => cell.trim()));
}

export default function TableCard({ artifact }) {
  const rows = useMemo(() => {
    if (Array.isArray(artifact?.rows)) {
      return artifact.rows;
    }
    if (typeof artifact?.csv === 'string') {
      return parseCsv(artifact.csv);
    }
    return [];
  }, [artifact]);

  if (!rows.length) {
    return <div className="text-sm text-slate-500">No table data available.</div>;
  }

  const [header, ...body] = rows;

  return (
    <div className="max-h-[520px] overflow-auto rounded border border-slate-200">
      <table className="min-w-full divide-y divide-slate-200 text-sm">
        <thead className="bg-slate-50">
          <tr>
            {header.map((cell, index) => (
              <th key={index} scope="col" className="px-4 py-2 text-left font-semibold text-slate-700">
                {cell}
              </th>
            ))}
          </tr>
        </thead>
        <tbody className="divide-y divide-slate-100 bg-white">
          {body.map((row, rowIndex) => (
            <tr key={rowIndex} className="hover:bg-slate-50">
              {row.map((cell, cellIndex) => (
                <td key={cellIndex} className="px-4 py-2 text-slate-600">
                  {cell}
                </td>
              ))}
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
}
