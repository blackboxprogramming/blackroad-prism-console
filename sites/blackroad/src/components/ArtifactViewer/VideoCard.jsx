import React from 'react';

export default function VideoCard({ artifact }) {
  if (!artifact?.url) {
    return <div className="text-sm text-slate-500">Video unavailable.</div>;
  }
  return (
    <div className="flex flex-col gap-3">
      <video controls className="max-h-[480px] w-full rounded border border-slate-200 shadow">
        <source src={artifact.url} type="video/webm" />
        Your browser does not support the WebM format.
      </video>
      {artifact.caption ? <p className="text-xs text-slate-500">{artifact.caption}</p> : null}
    </div>
  );
}
