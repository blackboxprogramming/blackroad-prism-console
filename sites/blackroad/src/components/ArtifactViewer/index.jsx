import React, { useCallback, useEffect, useMemo, useState } from 'react';
import ImageCard from './ImageCard.jsx';
import TableCard from './TableCard.jsx';
import VideoCard from './VideoCard.jsx';
import JsonCard from './JsonCard.jsx';
import DownloadButton from './DownloadButton.jsx';

const CARD_BY_KIND = {
  image: ImageCard,
  table: TableCard,
  video: VideoCard,
  json: JsonCard,
};

const DEFAULT_ARTIFACTS = [];

async function fetchArtifacts(jobId, signal) {
  const response = await fetch(`/api/jobs/${jobId}/artifacts`, { signal });
  if (!response.ok) {
    throw new Error(`Failed to load artifacts (${response.status})`);
  }
  return response.json();
}

export default function ArtifactViewer({ jobId, artifacts: initialArtifacts, onSelect }) {
  const [artifacts, setArtifacts] = useState(initialArtifacts ?? DEFAULT_ARTIFACTS);
  const [activeIndex, setActiveIndex] = useState(0);
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    if (!jobId || initialArtifacts) {
      return undefined;
    }
    const controller = new AbortController();
    setLoading(true);
    fetchArtifacts(jobId, controller.signal)
      .then((data) => {
        setArtifacts(Array.isArray(data) ? data : []);
        setActiveIndex(0);
        setError(null);
      })
      .catch((err) => {
        if (err.name !== 'AbortError') {
          setError(err);
        }
      })
      .finally(() => setLoading(false));
    return () => controller.abort();
  }, [jobId, initialArtifacts]);

  useEffect(() => {
    if (initialArtifacts) {
      setArtifacts(initialArtifacts);
      setActiveIndex(0);
    }
  }, [initialArtifacts]);

  useEffect(() => {
    const handler = (event) => {
      if (!artifacts.length) return;
      if (event.key === 'ArrowRight') {
        event.preventDefault();
        setActiveIndex((index) => (index + 1) % artifacts.length);
      }
      if (event.key === 'ArrowLeft') {
        event.preventDefault();
        setActiveIndex((index) => (index - 1 + artifacts.length) % artifacts.length);
      }
      if (event.key.toLowerCase() === 'd') {
        event.preventDefault();
        const artifact = artifacts[activeIndex];
        if (artifact?.downloadUrl) {
          window.open(artifact.downloadUrl, '_blank');
        }
      }
    };
    window.addEventListener('keydown', handler);
    return () => window.removeEventListener('keydown', handler);
  }, [artifacts, activeIndex]);

  const activeArtifact = artifacts[activeIndex];
  const Card = useMemo(() => (activeArtifact ? CARD_BY_KIND[activeArtifact.kind] ?? JsonCard : null), [activeArtifact]);

  useEffect(() => {
    if (activeArtifact && onSelect) {
      onSelect(activeArtifact, activeIndex);
    }
  }, [activeArtifact, activeIndex, onSelect]);

  const handleSelect = useCallback(
    (index) => {
      setActiveIndex(index);
    },
    [setActiveIndex],
  );

  return (
    <div className="flex h-full flex-col rounded-lg border border-slate-200 bg-white shadow-sm">
      <div className="flex items-center justify-between border-b border-slate-200 px-4 py-2">
        <div>
          <h2 className="text-sm font-semibold text-slate-900">Artifacts</h2>
          {jobId ? <p className="text-xs text-slate-500">Job {jobId}</p> : <p className="text-xs text-slate-500">Select a job</p>}
        </div>
        {activeArtifact?.downloadUrl ? <DownloadButton url={activeArtifact.downloadUrl} /> : null}
      </div>

      {error ? (
        <div className="p-4 text-sm text-red-600">{error.message}</div>
      ) : loading ? (
        <div className="p-4 text-sm text-slate-500">Loading artifactsâ€¦</div>
      ) : !artifacts.length ? (
        <div className="p-4 text-sm text-slate-500">No artifacts yet.</div>
      ) : (
        <div className="flex flex-1 overflow-hidden">
          <aside className="w-48 border-r border-slate-200 bg-slate-50">
            <ul className="max-h-full overflow-auto text-sm">
              {artifacts.map((artifact, index) => (
                <li key={artifact.id ?? index}>
                  <button
                    type="button"
                    onClick={() => handleSelect(index)}
                    className={`flex w-full items-center justify-between px-3 py-2 text-left hover:bg-slate-100 ${
                      index === activeIndex ? 'bg-slate-200 font-semibold' : 'text-slate-700'
                    }`}
                  >
                    <span>{artifact.name ?? artifact.kind}</span>
                    <span className="text-xs uppercase text-slate-500">{artifact.kind}</span>
                  </button>
                </li>
              ))}
            </ul>
          </aside>
          <div className="flex-1 overflow-auto p-4">
            {Card ? <Card artifact={activeArtifact} /> : <div className="text-sm text-slate-500">Unsupported artifact.</div>}
          </div>
        </div>
      )}
    </div>
  );
}
