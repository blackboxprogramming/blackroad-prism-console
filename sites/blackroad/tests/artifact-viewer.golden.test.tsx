import { describe, expect, it } from 'vitest';
import { render } from '@testing-library/react';
import ArtifactViewer from '../src/components/ArtifactViewer/index.jsx';

const ARTIFACTS = [
  {
    id: 'art-1',
    kind: 'image',
    name: 'Frame 1',
    url: '/static/frame1.png',
    downloadUrl: '/static/frame1.png',
  },
  {
    id: 'art-2',
    kind: 'table',
    name: 'Metrics',
    rows: [
      ['Metric', 'Value'],
      ['Loss', '0.018'],
      ['Drift', '0.002'],
    ],
  },
  {
    id: 'art-3',
    kind: 'json',
    name: 'Config',
    json: { learningRate: 0.001, schedule: 'cosine' },
  },
];

describe('ArtifactViewer', () => {
  it('renders deterministic artifact layout', () => {
    const { container } = render(<ArtifactViewer jobId="demo" artifacts={ARTIFACTS} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
