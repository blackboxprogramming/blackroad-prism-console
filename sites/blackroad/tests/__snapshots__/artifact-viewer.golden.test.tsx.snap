// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ArtifactViewer > renders deterministic artifact layout 1`] = `
<div
  class="flex h-full flex-col rounded-lg border border-slate-200 bg-white shadow-sm"
>
  <div
    class="flex items-center justify-between border-b border-slate-200 px-4 py-2"
  >
    <div>
      <h2
        class="text-sm font-semibold text-slate-900"
      >
        Artifacts
      </h2>
      <p
        class="text-xs text-slate-500"
      >
        Job 
        demo
      </p>
    </div>
    <a
      class="rounded border border-slate-300 bg-white px-3 py-1 text-xs font-medium text-slate-700 shadow hover:bg-slate-100"
      href="/static/frame1.png"
      rel="noreferrer"
      target="_blank"
    >
      Download
    </a>
  </div>
  <div
    class="flex flex-1 overflow-hidden"
  >
    <aside
      class="w-48 border-r border-slate-200 bg-slate-50"
    >
      <ul
        class="max-h-full overflow-auto text-sm"
      >
        <li>
          <button
            class="flex w-full items-center justify-between px-3 py-2 text-left hover:bg-slate-100 bg-slate-200 font-semibold"
            type="button"
          >
            <span>
              Frame 1
            </span>
            <span
              class="text-xs uppercase text-slate-500"
            >
              image
            </span>
          </button>
        </li>
        <li>
          <button
            class="flex w-full items-center justify-between px-3 py-2 text-left hover:bg-slate-100 text-slate-700"
            type="button"
          >
            <span>
              Metrics
            </span>
            <span
              class="text-xs uppercase text-slate-500"
            >
              table
            </span>
          </button>
        </li>
        <li>
          <button
            class="flex w-full items-center justify-between px-3 py-2 text-left hover:bg-slate-100 text-slate-700"
            type="button"
          >
            <span>
              Config
            </span>
            <span
              class="text-xs uppercase text-slate-500"
            >
              json
            </span>
          </button>
        </li>
      </ul>
    </aside>
    <div
      class="flex-1 overflow-auto p-4"
    >
      <figure
        class="flex h-full w-full flex-col items-center justify-center gap-3"
      >
        <img
          alt="Frame 1"
          class="max-h-[480px] max-w-full rounded border border-slate-200 object-contain shadow"
          src="/static/frame1.png"
        />
      </figure>
    </div>
  </div>
</div>
`;
